File: interpretado/testes.py
from Node import Node
from pilhaArray import PilhaArray
from filaArray import FilaArray
from filaListaEncadeada import FilaListaEncadeada
from pilhaListaEncadeada import PilhaListaEncadeada
import time
import tracemalloc
import random
import sys


def testarPilhaArray(n):
    print(f"\n[teste com {n} elementos - Pilha com Array]")

    pilha = PilhaArray(n)

    # teste do push
    inicio = time.time()
    for _ in range(n):
        pilha.push(random.randint(0, 1000))
    fim = time.time()
    tempoInsercao = fim - inicio

    #t teste do pop
    inicio = time.time()
    while not pilha.isEmpty():
        pilha.pop()
    fim = time.time()
    tempoRemocao = fim - inicio

    # memória estimada
    memoria_mb = (sys.getsizeof(pilha.arr) + sys.getsizeof(pilha)) / (1024 * 1024)

    print(f"> tempo de insercao: {tempoInsercao:.6f} segundos")
    print(f"> tempo de remocao: {tempoRemocao:.6f} segundos")
    print(f"> memoria estimada: {memoria_mb:.2f} MB")


def testarPilhaEncadeada(n):
    print(f"\n[teste com {n} elementos - Pilha com Lista Encadeada]")

    pilha = PilhaListaEncadeada()

    #testando o push
    tracemalloc.start()
    inicio_insercao = time.time()
    for i in range(n):
        pilha.push(i)
    fim_insercao = time.time()
    tempo_insercao = fim_insercao - inicio_insercao

    #testando o pop
    inicio_remocao = time.time()
    for i in range(n):
        pilha.pop()
    fim_remocao = time.time()
    tempo_remocao = fim_remocao - inicio_remocao

    #memoria estimada
    memoria_usada = tracemalloc.get_traced_memory()[1] / (1024 * 1024)
    tracemalloc.stop()

    print(f"> tempo de insercao: {tempo_insercao:.6f} segundos")
    print(f"> tempo de remocao:  {tempo_remocao:.6f} segundos")
    print(f"> memoria estimada:  {memoria_usada:.2f}MB")

def testarFilaArray(n):
    print(f"\n\n[teste com {n} elementos - Fila com Array estatico]")

    fila = FilaArray(capacity=n + 1) 

    tracemalloc.start()
    inicio_insercao = time.time()
    #testando o enqueue
    for _ in range(n):
        fila.enqueue(random.randint(0, 1000))

    fim_insercao = time.time()
    tempo_insercao = fim_insercao - inicio_insercao

    
    inicio_remocao = time.time()
    #testando o dequeue
    while not fila.isEmpty():
        fila.dequeue()
    fim_remocao = time.time()
    tempo_remocao = fim_remocao - inicio_remocao

    memoria_usada, _ = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    #resultados
    print(f"> tempo de insercao: {tempo_insercao:.6f} segundos")
    print(f"> tempo de remocao: {tempo_remocao:.6f} segundos")
    print(f"> memoria estimada: {memoria_usada / (1024 * 1024):.2f} MB")


def testarFilaEncadeada(n):
    print(f"\n[teste com {n} elementos - Fila com Lista Encadeada]")

    fila = FilaListaEncadeada()

    #testando o enqueue
    inicio_insercao = time.time()
    for _ in range(n):
        fila.enqueue(random.randint(0, 1000))
    fim_insercao = time.time()
    tempo_insercao = fim_insercao - inicio_insercao

    #testando o dequeue
    inicio_remocao = time.time()
    while not fila.isEmpty():
        fila.dequeue()
    fim_remocao = time.time()
    tempo_remocao = fim_remocao - inicio_remocao

    # estimativa de memoria ajustada pois o tracemalloc nao estava pegando a memoria usada
    memoria_bytes = n * 64 + 64
    memoria_mb = memoria_bytes / (1024 * 1024)

    #resultados
    print(f"> tempo de insercao: {tempo_insercao:.6f} segundos")
    print(f"> tempo de remocao: {tempo_remocao:.6f} segundos")
    print(f"> memoria estimada: {memoria_mb:.2f} MB")

File: compilado/src/testes.c
#include "../includes/libs.h"
#include "../includes/filaArray.h"
#include "../includes/pilhaArray.h"
#include "../includes/filaListaEncadeada.h"
#include "../includes/pilhaListaEncadeada.h"
#define MAX 1500000


void testarFilaArray(int n) {
    printf("\n[Fila com array estatico]\n");

    FilaArray fila;
    criarFilaArray(&fila);

    clock_t inicio, fim;
    double tempoInsercao, tempoRemocao;

    //testando o enqueue
    inicio = clock();
    for (int i = 0; i < n; i++) {
        enqueueFilaArray(&fila, rand() % 1000);
    }
    fim = clock();
    tempoInsercao = (double)(fim - inicio) / CLOCKS_PER_SEC;

    //testando o dequeue
    inicio = clock();
    while (!isEmptyFilaArray(&fila)) {
        dequeueFilaArray(&fila);
    }
    fim = clock();
    tempoRemocao = (double)(fim - inicio) / CLOCKS_PER_SEC;

    // Estimativa de memória
    size_t bytes = sizeof(int) * MAX_SIZE;
    double memoriaMB = bytes / (1024.0 * 1024.0);

    printf("> tempo de insercao: %.6f segundos\n", tempoInsercao);
    printf("> tempo de remocao: %.6f segundos\n", tempoRemocao);
    printf("> memoria estimada: %.2fMB\n", memoriaMB);
}

void testarPilhaArray(int quantidade) {
    printf("\n[Pilha com array estatico]\n");

    PilhaArray pilha;
    pilha.header = 0;

    clock_t inicio, fim;
    double tempoInsercao, tempoRemocao;

    //teste com o push
    inicio = clock();
    for (int i = 0; i < quantidade; i++) {
        pushPilhaArray(&pilha, rand() % 1000);
    }
    fim = clock();
    tempoInsercao = (double)(fim - inicio) / CLOCKS_PER_SEC;

    //teste com o pop
    inicio = clock();
    while (pilha.header > 0) {
        popPilhaArray(&pilha);
    }
    fim = clock();
    tempoRemocao = (double)(fim - inicio) / CLOCKS_PER_SEC;

    //memoria estimada
    double memoriaMB = (sizeof(PilhaArray)) / (1024.0 * 1024.0);

    printf("> Tempo de insercao: %.6f segundos\n", tempoInsercao);
    printf("> Tempo de remocao: %.6f segundos\n", tempoRemocao);
    printf("> memoria estimada: %.2fMB\n", memoriaMB);
}


void testarFilaEncadeada(int quantidade) {
    printf("\n[Fila com lista encadeada]\n");

    FilaEncadeada* fila = criarFilaEncadeada();

    clock_t inicio, fim;
    double tempoInsercao, tempoRemocao;

    //teste do enqueue
    inicio = clock();
    for (int i = 0; i < quantidade; i++) {
        enqueueFilaEncadeada(fila, rand() % 1000);
    }
    fim = clock();
    tempoInsercao = (double)(fim - inicio) / CLOCKS_PER_SEC;

    //teste do dequeue
    inicio = clock();
    while (!isEmptyFilaEncadeada(fila)) {
        dequeueFilaEncadeada(fila);
    }
    fim = clock();
    tempoRemocao = (double)(fim - inicio) / CLOCKS_PER_SEC;

    //memoria estimada: (tamanho de cada node * quantidade) + tamanho da estrutura
    double memoriaMB = ((sizeof(No) * quantidade) + sizeof(FilaEncadeada)) / (1024.0 * 1024.0);

    printf("> tempo de inserção: %.6f segundos\n", tempoInsercao);
    printf("> tempo de remoção: %.6f segundos\n", tempoRemocao);
    printf("> memoria estimada: %.2fMB\n", memoriaMB);

    free(fila); //libera a estrutura
}


void testarPilhaEncadeada(int qtdElementos) {
    printf("\n[Pilha com Lista Encadeada]\n");

    PilhaEncadeada* pilha = createPilhaEncadeada();

    //testanto push
    clock_t inicioInsercao = clock();
    for (int i = 0; i < qtdElementos; i++) {
        pushPilhaEncadeada(pilha, i);
    }
    clock_t fimInsercao = clock();
    double tempoInsercao = (double)(fimInsercao - inicioInsercao) / CLOCKS_PER_SEC;

    //testando o pop
    clock_t inicioRemocao = clock();
    for (int i = 0; i < qtdElementos; i++) {
        popPilhaEncadeada(pilha);
    }
    clock_t fimRemocao = clock();
    double tempoRemocao = (double)(fimRemocao - inicioRemocao) / CLOCKS_PER_SEC;

    // estimativa de memoria
    double memoriaBytes = sizeof(PilhaEncadeada) + qtdElementos * sizeof(No);
    double memoriaMB = memoriaBytes / (1024 * 1024);

    // resultados
    printf("> tempo de insercao: %.6f segundos\n", tempoInsercao);
    printf("> tempo de remocao: %.6f segundos\n", tempoRemocao);
    printf("> memoria estimada: %.2fMB\n", memoriaMB);

    // limpeza
    freePilhaEncadeada(pilha);
}